Merged discussions to RFC1004 (#24)
edgedb/rfcs: text/1004-transactions-api.rst

`More discussion on this topic <https://github.com/edgedb/edgedb/discussions/1708>`_

> This part of API change is also motivated by the fact that PostgreSQL

Author: My point was that unprepared reader can read this and say that "edgedb is less stable and throws more errors than postgres and that all this transaction repeating thing is a workaround to handle this", which isn't true. Another backend could tackle this issue differently (i.e. locking everything). And we generally don't hide that we're using postgres as a backend.

With all that said should I replace Postgres to EdgeDB?

> This part of API change is also motivated by the fact that PostgreSQL

Reviewer: I think writing it as "EdgeDB (as is PostgreSQL)" strikes a good balance between your and Yury's concern.

    class Pool(ReadOnlyPool):
        _impl: _PoolImpl
        _desired_config: SessionConfig

    class ReadOnlyPool:
        _impl: _SingletonPool
        _desired_config: SessionConfig

    class _PoolImpl:
        _connections: deque[RawConnection]  # Simplified

Both ``Pool`` and ``Connection`` have:

1. ``execute``, ``query`, ``query_one`` set of query functions
   (implement ``Executor`` abstract class defined below). All the
   methods reconnect if called on broken connection. Read-only queries
   broken in the middle of the query are retried.
2. ``retry`` method (described before)
3. ``try_transaction`` method for more fine-grained control over
   transactions (this is older ``transaction`` method, see below)

``ReadOnly*`` counterparts differ from non-read-only ones in two
important ways:

1. They send read-only flag to the compiler, so compiler rejects queries
   that modify data. And *may* use replica connection.
2. Works as type-checker hint that function that received a connection
   does no modifications to the data.

``RawConnection`` class is exposed for users want to do their own
session config management stuff, connection pooling or control of
reconnection.

> Note: this section contains only sync Python example, async Python and

Author: Oh, well, forgot that entirely. My idea was that `AsyncIO` prefixes and `async` in signatures don't add any value, but...

> Note: this section contains only sync Python example, async Python and

Author: Fixed

JavaScript bindings undergo changes similar enough that we don't think
it makes sense to put them here explicitly.

Connection Configuration Methods
--------------------------------

The following things are modified by a method that returns a distinct
object of the same type and same underlying network connections but
distinct config (i.e. ``Connection`` retains the same
``_SingletonPool`` and ``Pool`` retains the same ``_PoolImpl``):

* Asking for a read-only access
* Setting session configuration
* Transaction options

Here are method signatures:

.. code-block:: python

    class Connection:
        ...

>         def read_only(self,

Reviewer 2: I think `as_*` is more common for this kind of API.

>         def read_only(self,

Author: Then we have to change all other methods. But `as_session_config` is not the same as `with_session_config`. So let's make `with_read_only` then?

        def with_session_config(self, **config) -> Connection: ...
        def with_transaction_options(self, isolation: ...) -> Connection: ...
        def with_retry_options(self, attempts: int = 3, ...) -> Connection: ...

    class Pool:
        ...
        def read_only(self,
            primary: bool = false,
            allow_upgrade: bool = false,  # allows with_modifications
        ) -> ReadOnlyConnection: ...
        def with_modifications(self) -> Connection: ...
        def with_session_config(self, **config) -> Pool: ...
        def with_transaction_options(self, isolation: ...) -> Pool: ...
        def with_retry_options(self, attempts: int = 3, ...) -> Pool: ...

The ``ReadOnlyPool`` and ``ReadOnlyConnection`` get the same methods
(including ``read_only`` method itself, which is no-op).

After modification, connection/pool object can be used interchangeably:

.. code-block:: python

    conn = edgedb.connect()
    read_only_conn = conn.read_only()
    conn.execute("INSERT User { ... }", ...)
    print(read_only_conn.query("SELECT User"))
    read_only_conn.execute("INSERT User { .. }", ...)  # throws an error
    conn.read_only(allow_upgrade=True) \

>         .with_modifications().execute("INSERT User { ... }", ...)  # good

Author: Well, I still think this is very useful pattern that I would use in **every** my project. But if we have to cut it to do at least one step forward, okay.

``EarlyNetworkError`` or ``NetworkError``. Former is used in context
where we catch network error before sending a request.

> ``TransactionActiveError`` is introduced to signal that

Author: `TransactionIsActiveError`? 

`ActiveTransactionError` -- reads as error of active transaction (i.e. some base for a whole class of errors).

Also added clarification below.

> queries can't be executed on the connection object:

Author: Added explanation

Additionally ``TransactionActiveError`` is introduced to signal that
queries can't be executed on the connection object:

.. code-block:: python

    class TransactionActiveError(InterfaceError):
        _code = 0x_FF_02_01_04

Server Availability Timeout
---------------------------

Previously, when TCP connect gets "connection refused" error or when
timeout happens on handshake both connection and pool API would crash

> the application. This is weird when initially starting an application

Author: Changed to "can be inconvenient". This isn't particularly confusing.

cluster simultaneously with the database, or just when starting a
projece locally using ``docker-compose up``.

This RFC introduces a connection parameter:

.. code-block:: python

>     edgedb.connect('inst1', wait_until_available_sec=30)

Reviewer 2: Hard and solid -1 from me. One library shouldn't fight the language for unclear benefits.

2. If first connection could not be established for this timeout, error
   is thrown that holds the reason of the last failure.
3. On the subsequent reconnects this timeout is also obeyed (i.e.
   connection might try to reconnect multiple times) and error is thrown
   after the specified timeout as a result of the operation that
   requires connection (i.e. on query or transaction start).
4. For ``retry`` we bail out if this timeout is reached during any
   single reconnect. We don't retry immediately after database is marked
   as unavailable.
5. Subsequent queries or transactions after failure will retry for the
   specified timeout each time.

Only the following conditions are treated as eligible for reconnect:

1. Name resolution failed
2. File not found (Unix socket not bound yet)
3. Connection reset, connection aborted, connection refused (server is
   restarting or not ready yet)
4. Timeout happened during connect or authentication

All other errors are propagated immediately.

This timeout is different from ``connect_timeout``. Connect timeout
determines how long individual connect attempt may take. And
``wait_until_available`` determines how many such attempts could be

> made (i.e. how many ones fits the time frame).

Author: Previously we discussed that at least with @1st1 and he was for this change.

Here are my points:
1. For nice `docker-compose up` we need `wait_until_available` = 30 sec
2. Current python timeout is 60 sec. So by introducing `wait_until_available` 30 sec we don't make application hanging longer than what was there. In particular, I propose to lower `connect_timeout` to 10 seconds (for a single connection attempt). With connection pooling (ports RFC) this should work well enough.
3. If default is `0` nobody would ever discover `wait_until_available` and will put their own hacks for running locally, like they do with postgres. Unless we put this setting in every tutorial (which we definitely don't want for conciseness).
4. We can't crash app on every `query()` that got reconnected and caught `ConnectionRefused` (and get app restarted). So it's also a bunch of 500 errors on EdgeDB restart or failover.

*(the case when app and database starts simultaneously with the database on kubernetes is so much rare that we don't care about that use case)*

EdgeDB Changes
--------------

To support features above we add two headers to EdgeDB queries:

1. For PrepareComplete_, CommandComplete_ server-side messages:

>    ``SERVER_OPT_HAS_FEATURES``

Author: It's kinda just symmetry with ``QUERY_OPT``. ``SERVER_HAS_FEATURES`` reads bad. But better name is appreciated.

2. For Prepare_, OptimisticExecute_, Execute_, ExecuteScript_
   client-side messages: ``QUERY_OPT_ALLOW_FEATURES``

Both contain the set of strings joined by comma:

1. `modification` -- query is not read-only

> 2. `session` -- query contains session config change

Author: Changed and also use bitmap, so names matter much less.

3. `config` -- server or database config change
4. `ddl` -- query contains DDL
5. `transaction` -- query contains start/commit/rollback of transaction

>     or savepoint manipulation

Author: I don't think that performance here would be noticeable, but bitmap sounds good anyway. and using all ones to "allow everything" is good too.

In case of ``SERVER_OPT_HAS_FEATURES`` it describes what is actually
contained in the query. And in case of ``QUERY_OPT_ALLOW_FEATURES``

> client can specify what of these things are allowed in this query.

Author: Not sure I understand. It's in the next paragraph.

Read-only queries are always allowed. When ``QUERY_OPT_ALLOW_FEATURES``
is omitted any query is allowed (default). With the empty string only

> read-only queries are allowed.

Reviewer 2: I don't understand why we need this complexity.  IMO instead of "modification" header flag we should have "readonly". If it's set the query fails to prepare unless it's readonly. This is way easier to understand and document.

> read-only queries are allowed.

Author: This is because:
```
conn.execute("""
START TRANSACTION;
UPDATE User SET {...}
COMMIT
""")
```
Would return `transaction,modifications` flag.

With @1st1 proposal `.execute("START TRANSACTION")` would return `transaction,readonly` features and `.execute("START TRANSACTION; UPDATE..."` would return `transaction`.

I.e. purely additive features are better at least for `HAS_FEATURES`, but is not bad for `ALLOW` too.

I we go with bitmask read-only queries == '0' is also good enough.

> read-only queries are allowed.

Reviewer: I agree with @tailhook.  It's easier to reason when each bit _extends_ capabilities consistently as opposed to either restricting or extending them.

``SERVER_OPT_HAS_FEATURES`` is an empty string for read-only queries (the
field is present) as it indicates that query has been analyzed.

The ``SERVER_OPT_HAS_FEATURES`` is needed for the following tasks:

> 1. Retry standalone (non-transactional) read-only queries

Reviewer 2: You mean that the server would retry some of the client's queries on its own? I wouldn't like that.

Author: Not server, client.

2. Warn when features are used in inapropriate context (warn when
   session modification queries are sent on non-raw connection, which

>    means they can be lost at any point due to reconnect)

Reviewer 2: Hm, how will the server know if the connection is raw or non-raw? IMO it shouldn't care, it's entirely up to the client.

Author: Here is the strawman (but all of these are implementation details):
```
class RawConnection:
    def _query_with_options(self, options, query, *args, **kw): ...
    def query(self, query, *args, **kw):
        self._query_with_options({'allow': '*'}, query, *args, **kw)

class Connection:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'warn': ['transaction','session_config']}, query, *args, **kw)

class ReadonlyConnection:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'allow': []}, query, *args, **kw)

class PoolConnectionProxy:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'deny': ['transaction','session_config']}, query, *args, **kw)

class Transaction:
    def __enter__(self):
          self._conn._raw._query_with_options({'allow': ['transaction']}, "START TRANSACTION")
    def query(self, *args, **kw):
          self._conn._raw._query_with_options({'allow': ['modifications', 'ddl']}, query, *args, **kw)
````
(these allow/deny are to show the concept, we're probably going to have a constant with the bit mask already prepared for the specific use case for efficiency)

3. Connection warns on ``session`` and ``transaction``
4. Connection got from pool errors on ``session`` and ``transaction``
5. Everything is allowed in ``RawConnection``

Note: session settings and transactions should be activated using
special methods ``with_session_config``, ``try_transaction`` and
``retry``, rather than by using ``execute(...)`` or ``query(...)`` in
the former case they are allowed internally. And this should be
indicated in the respective error messages.

.. _PrepareComplete: https://www.edgedb.com/docs/internals/protocol/messages#preparecomplete
.. _CommandComplete: https://www.edgedb.com/docs/internals/protocol/messages#commandcomplete
.. _Prepare: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-prepare
.. _OptimisticExecute: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-optimistic-execute
.. _Execute: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-execute
.. _ExecuteScript: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-execute-script

Future Work
===========

More Configuration
------------------

> Setting tracing/debugging metadata for queries may be implemented

Author: That is a future work. Tags are essentially a specific way to organize metadata. Also OpenTracing is deprecated, and OpenTelemetry should be used instead.

Also `with_tags` has little meaning. `with_telemetry_tags` or `with_tracing_tags` looks better, but let's not bikeshed about this here.

> in the same way:

Reviewer 2: ```suggestion
in the same way in the future:
```

.. code-block:: python

    def handle(request, db):
        db = db.with_metadata(
            uri=request.uri,
            username=request.session.get('username'),
        )
        handle_user_request(request, db)

And this can be done on transaction level too:

.. code-block:: python

    async for tx in db.retry():
      async with tx:
        tx = tx.with_metadata(
            uri=request.uri,
            username=request.session.get('username'),
        )
        tx.query("...")

More Read-Only Options
----------------------

Replica config may be specified when configuring a read-only connection:

.. code-block:: python

   def read_only(self, primary=False, max_replica_lag=10): ...

> Perhaps this should be encapsulated into replica options:

Reviewer 2: Let's not use "perhaps" here -- we need to decide now or leave this for future RFCs

Author: This is a "Future Work" section. What's wrong in "perhaps" here? This just gives more context to the current decisions but don't have to be clearly defined in this RFC.

.. code-block:: python

   def read_only(self, primary=False, replicas: ReplicaOptions): ...

Learning Curve
==============

> The ``retry`` method complicates learning curve curve, but:

Reviewer 2: ```suggestion
The ``retry`` method complicates the learning curve, but:
```

1. This is already a problem in Postgres, there a lot of people who
   ignore the issue for pet projects and a lot of startups having
   "normal" rate or 500 errors at any point of time, while it's

>    preventable.

Reviewer 2: "1. Letting the application to error out instead of automatically retrying certain transactions is a wildly known, yet an entirely preventable problem."

2. Repeating transactions would make less incentive to keep transactions
   open for a long time (e.g. while accessing slow network resources

>    like external API), which is a problem of itself.

Reviewer 2: "2. The design of the proposed `retry()` method emphasizes that the code block might be executed more than one time, suggesting to the user to factor out slow blocking code, like making API calls over network. This ensures that DB transactions would not be open longer than it is necessary."

3. Even if we never have failed concurrent updates we would want
   seamless reconnect on connection failures (i.e. server restart,

>    primary/replica change, etc.)

Reviewer 2: "3. The API adds resilience against not only errors related to failing to serialize concurrent updates, but also against possible network errors caused by the DB server restart, primary/replica changes, etc."

4. To make learning curves shorter I think we should intentionally
   inject failures. This is needed so that users quickly find out that

>    side effects of their transactions are in effect several times.

Reviewer 2: "4. In the future we will consider adding a connection setting to inject failures into transactions to help users discover incorrect transactional code."

@tailhook @elprans I think we shouldn't do this before 1.0 as I'm not sure how it might impact benchmarks, production code, and the overall DX of early adopters. I think that our API is already going to be structured in the right way, and adding failure injection is entirely optional at this point.

Reviewer: I also think that proper API design, documentation and tutorials would be sufficient to guide users toward the correct set of idioms.  Intentional failure injection might be an optional feature.

Author: This is where we constantly disagree. Documentation and tutorials are not enough for the same reason for which we use unit tests to test code. As we now basically know the distinction between workstation installations (created by `edgedb server init`) and production (roughly -- everything else), it's possible to have server itself without failure injection by default, but `server init` configuring it in a way that enables this by default. Also that section describes a way of failure injection that doesn't impact benchmarks in any significant way.

But anyway, this RFC is already hanging for too long, so I drop the "failure injection" section in favor of later RFC.

2. For Prepare_, OptimisticExecute_, Execute_, ExecuteScript_
   client-side messages: ``QUERY_OPT_ALLOW_FEATURES``

> Both contain 32bit bitmap of the following:

Reviewer: Let's do 64 bits.  No reason to conserve space here.

   network failures would be an issue

> Keep `transaction` as is but Add a Helper for Retrying

Reviewer 2: ```suggestion
Separate Retry API
```

------------------------------------------------------

> The problem with this approach is that it hard to teach using ``retry``

Reviewer 2: "The problem of adding a separate *transaction retry API* while keeping the current `Connection.transaction()` family of methods is that it will be hard to educate the users that they should almost always use the former. Most of the time transaction serialization and network errors are uncommon during the development and only surface when in production."

when raw transactions "work on my laptop". However, this is somewhat
alleviated by failure injection.

Retry All Single Queries
------------------------

This specification describes that read-only non-transactional queries
should be retried automatically.

While it's tempting to retry all ``pool.query`` and ``pool.execute``

> calls (even modifications), it **gives the false sense of security**: no

Reviewer 2: ```suggestion
calls (even those that execute queries with side effects), it **gives the false sense of security**: no
```

"concurrent update" issues are seen.  But it's better to see such error
and turn the whole block of code into a transaction rather than just a
mutation. I.e. retrying a single mutable request on a "concurrent

> update" error must be a deliberate decision.

Reviewer 2: This whole paragraph is very hard to parse. I suggest to reword it to:

"Automatically retrying `query()`-like calls that execute queries with side effects is not always safe. Therefore our recommendation to users would be to use the new `retry()` API when they know that the query would be safe to repeat."

Author: This is not enough in my opinion. Added an example for clarification.

Is ``with_modifications`` Needed?
---------------------------------

It's intuitive that databases are mutable by default. But there is a
large class of applications that are mostly read-only and must have
limited and easy to find places having mutations. For those apps it's
better to have read-only connection by default and use a pattern like

> this for writes:

Reviewer 2: The RFC starts with stating and showing examples of `edgedb.connect()` returning a mutable connection. So I assume that this section advocates for adding a way to convert a read-only connection (obtained via `edgedb.connect().read_only()`) to a mutating one?  I'm confused.

Reviewer 2: Yeah. Paul, we've argued over this a lot already, and I see no end to this. So I propose we do this instead: we don't implement `with_modifications()` right now. We can always add this later, it's really not a critical feature and we can add it at any time when we get some user feedback.

I'd really like this RFC to only implement essential and **non-controversial** APIs, where non-controversial means unanimous support of you, Elvis, and I.

Author: > So I assume that this section advocates for adding a way to convert a read-only connection (obtained via edgedb.connect().read_only()) to a mutating one?

Yes, in the top of the application, you either do (assuming `aiohttp`):
```python
app['edgedb'] = await create_async_pool('name')
```
Or you do:
```python
app['edgedb'] = (await create_async_pool('name')).read_only(allow_upgrade=True)
```
In the latter case, you prevent arbitrary functions execute side-effects unless run within `with_modification().retry()` transaction and transaction is passed to the function directly.

> In my mind, the main feature and premise of read_only() is that it allows better scalability by allowing connections to standby replicas.

Well:
1. We don't implement that scalability just yet anyway
2. With regards of this RFC the original premise of read-only was retrying queries automatically. But well, it's solved by `HAS_FEATURES` so is not strictly needed
3. I don't see anything in `with_modifications` as controversial (except the name). Controversial would be official recommendation of this, but not the support. And in previous discussion I think it was agreed as "yes we can support this pattern".

Anyway I'm dropping the function and move this section to the Future Work, so that we can finally move forward.

  ``handler(c); handler(c)`` will not set session config second
  time, but use the knowledge that these settings are already active.

Also it improves clarity and composition. E.g. (here we demonstrate
``with_metadata`` method with is a part of `Future Work`_:

.. code-block:: python

    def money_transfer(tx):
        user1 = tx.with_metadata(user='user1')
        user2 = tx.with_metadata(user='user2')
        withdraw(user1)
        deposit(user2)
        log_withdraw(user1)
        log_deposit(user2)

(This implies ``Transaction`` also contains ``with_metadata`` method)

Additional thing considered is that with context manager, if connection
is failed in the middle of the block:

.. code-block:: python

    with conn.session_config(name="value"):
        conn.query("...")
        # << -- here
        conn.query("...")

The implementation should reconnect and reapply the session
configuration anyway. So it deviates from the usual behavior of context
manager anyway.

Disabling Features
------------------

We may introduce a pool and/or connection configuration to disable
``ddl`` and ``config`` features on the requests. And/or disable

> ``session`` and ``transaction`` features on connection. This should be

Reviewer 2: I like the disabling idea. I'm not sure if it should be the default configuration though, we can just teach that it's a good pattern to narrow down the connection functionality to what your app needs.

Reviewer: Agree.  I think we should optimize for simplicity at a glance, in which case lengthy connection "setups" in simple tutorials would scare users away.

Author: Looks like you agree with what is written in the RFC. Resolving the conversation.
